### PR TITLE
Support users and teams in branch protection access restrictions

### PR DESCRIPTION
**Pull request by jporzucek**
Fixes https://github.com/google/go-github/issues/2560

It also adds a ListAppRestrictions wrapper around ListApp so naming conventions with ListUserRestrictions and ListTeamRestrictions methods is preserved.

### CODE REVIEW

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1909]: slug => users

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @jporzucek !

File: github/repos.go
Comment (gmlewis) [reply github/repos.go:1884]: OK, but I think we can do better than the official documentation, don't you?

File: github/repos.go
Comment (gmlewis) [reply github/repos.go:1778]: Well, according to the linked documentation, this argument should really be called "teams" (possibly with a comment describing the argument as "The slug values for teams."
In fact, please go through all the added endpoints and make sure that the documentation and the parameter names make sense.  Please let me know when you feel they are appropriately named, and I'll start another full review. Thanks.

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1922]:
```suggestion
// RemoveUsersRestrictions removes the restrictions of a user from pushing to this branch.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1903]:
```suggestion
// Note: The list of users is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1881]:
```suggestion
// Note: The list of users is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1838]:
```suggestion
// Note: The list of teams is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1816]:
```suggestion
// Note: The list of teams is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1751]: // Note: The list of apps is limited to 100 items.

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1729]:
```suggestion
// Note: The list of apps is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1155]: Please revert this change.

ksnip_20230114-072357

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1157]: Please revert this change.

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1119]: Please revert this change.

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1121]: Please revert this change.

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1179]: Please revert this change.

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1181]:
```suggestion
	// The list of apps slugs which can dismiss pull request reviews. (Required; use nil to disable dismissal_restrictions or &[]string{} otherwise.)
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1707]:
```suggestion
// Note: The list of apps is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1794]:
```suggestion
// Note: The list of teams is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1877]:
```suggestion
// ReplaceUsersRestrictions replaces the user that have push access to a given protected branch.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1857]: Let's please change this to `ListUsersRestrictions` , since we are returning potentially many users and not a single user.

```suggestion
// ListUsersRestrictions lists the GitHub users that have push access to a given protected branch.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1900]:
```suggestion
// AddUsersRestrictions grants the specified users push access to a given protected branch.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1925]:
```suggestion
// Note: The list of users is limited to 100 items.
```

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1700]: Please change the existing unit test that tests `ListApps` to test `ListAppRestrictions` instead to make the Codecov reduction go away.

File: github/repos.go
Comment (gmlewis) [inline github/repos.go:1693]: Before line 1677 above, please add two new lines:

```go
//
// Deprecated: Please use ListAppRestrictions instead.
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thanks, @jporzucek .

File: github/repos.go
Comment (gmlewis) [reply github/repos.go:1903]: OK, gotcha.  Let's leave it then.

File: github/repos.go
Comment (gmlewis) [reply github/repos.go:1857]: I see.  OK, let's leave this as well, then.

**Review by gmlewis — APPROVED** [review]
Thank you, @jporzucek !
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Review by valbeat — APPROVED** [review]
LGTM!

**Pull request merged**
